lib/_stream_writable.js

```diff
@@ -497,67 +469,42 @@ function onwriteDrain(stream, state) {
 
 // If there's something in the buffer waiting, then process it
 function clearBuffer(stream, state) {
-  state.bufferProcessing = true;
-  var entry = state.bufferedRequest;
-
-  if (stream._writev && entry && entry.next) {
-    // Fast case, write everything using _writev()
-    var l = state.bufferedRequestCount;
-    var buffer = new Array(l);
-    var holder = state.corkedRequestsFree;
-    holder.entry = entry;
-
-    var count = 0;
-    var allBuffers = true;
-    while (entry) {
-      buffer[count] = entry;
-      if (!entry.isBuf)
-        allBuffers = false;
-      entry = entry.next;
-      count += 1;
-    }
-    buffer.allBuffers = allBuffers;
+  const buffered = state.buffered;
+  const bufferedLength = buffered.length;
 
-    doWrite(stream, state, true, state.length, buffer, '', holder.finish);
+  if (!bufferedLength || state.corked || state.bufferProcessing ||
+    state.writing) {
+    return;
+  }
 
-    // doWrite is almost always async, defer these to save a bit of time
-    // as the hot path ends with doWrite
+  state.bufferProcessing = true;
+  if (bufferedLength === 1) {
+    const { chunk, encoding, callback } = buffered[0];
+    const len = state.objectMode ? 1 : chunk.length;
+    doWrite(stream, state, false, len, chunk, encoding, callback);
+    buffered.length = 0;
+    buffered.allBuffers = true;
+  } else if (stream._writev) {
     state.pendingcb++;
-    state.lastBufferedRequest = null;
-    if (holder.next) {
-      state.corkedRequestsFree = holder.next;
-      holder.next = null;
-    } else {
-      var corkReq = { next: null, entry: null, finish: undefined };
-      corkReq.finish = onCorkedFinish.bind(undefined, corkReq, state);
-      state.corkedRequestsFree = corkReq;
-    }
-    state.bufferedRequestCount = 0;
-  } else {
-    // Slow case, write chunks one-by-one
-    while (entry) {
-      var chunk = entry.chunk;
-      var encoding = entry.encoding;
-      var cb = entry.callback;
-      var len = state.objectMode ? 1 : chunk.length;
-
-      doWrite(stream, state, false, len, chunk, encoding, cb);
-      entry = entry.next;
-      state.bufferedRequestCount--;
-      // If we didn't call the onwrite immediately, then
-      // it means that we need to wait until it does.
-      // also, that means that the chunk and cb are currently
-      // being processed, so move the buffer counter past them.
-      if (state.writing) {
-        break;
+    // doWrite mutates buffered array. Keep a copy of callbacks.
+    const callbacks = buffered.map(({ callback }) => callback);
+    doWrite(stream, state, true, state.length, buffered, '', (err) => {
+      for (const callback of callbacks) {
+        state.pendingcb--;
+        callback(err);
       }
+    });
+    state.buffered = [];
+    state.buffered.allBuffers = true;
+  } else {
+    let i = 0;
+    while (i < bufferedLength && !state.writing) {
```

Comment:
i think this could be turned into a for-loop?:

```js
for (let i = 0; i < bufferedLength && !state.writing; ++i) {
  const entry = buffered[i];
  // ...
}
```

Reply:
`i` needs to outlive the scope, but otherwise yes.